### PR TITLE
clp-package: Deduplicate and clean up code in CLI job-launcher scripts.

### PR DESCRIPTION
**Pull request by haiqi96**
# Description
This PR factors out the common code used in job submission scripts in clp_package_utils. It also move some query job related functions into a separate util file to prepare an upcoming change that supports ir extraction job in the decompression script.

# Validation performed
Launched the package, verified that compression, decompression and search still work from command line

### CODE REVIEW

File: components/clp-package-utils/clp_package_utils/general.py [hunk 1]
```diff
@@ -241,6 +257,43 @@ def generate_container_config(clp_config: CLPConfig, clp_home: pathlib.Path):
     return container_clp_config, docker_mounts
 
 
+def dump_container_config(
+    clp_config: CLPConfig, container_clp_config: CLPConfig, container_name: str
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/general.py:261]: Can we swap `container_clp_config` and `clp_config`?

File: components/clp-package-utils/clp_package_utils/general.py [hunk 2]
```diff
@@ -241,6 +257,43 @@ def generate_container_config(clp_config: CLPConfig, clp_home: pathlib.Path):
...
+    config_file_path_on_host = clp_config.logs_directory / container_config_filename
+    config_file_path_on_container = container_clp_config.logs_directory / container_config_filename
+    with open(config_file_path_on_host, "w") as f:
+        yaml.safe_dump(container_clp_config.dump_to_primitive_dict(), f)
+
+    return config_file_path_on_container, config_file_path_on_host
+
+
+def generate_container_start_cmd(
+    container_name: str, container_mounts: List[CLPDockerMounts], execution_container: str
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/general.py:273]: How about `container_image` instead of `execution_container`?

File: components/clp-package-utils/clp_package_utils/scripts/compress.py [hunk 3]
```diff
@@ -67,41 +67,23 @@ def main(argv):
...
 
-    container_name = f"clp-compressor-{str(uuid.uuid4())[-4:]}"
+    container_name = generate_container_name(JobType.COMPRESSION)
 
     container_clp_config, mounts = generate_container_config(clp_config, clp_home)
-    container_config_filename = f".{container_name}-config.yml"
-    container_config_file_path_on_host = clp_config.logs_directory / container_config_filename
-    with open(container_config_file_path_on_host, "w") as f:
-        yaml.safe_dump(container_clp_config.dump_to_primitive_dict(), f)
+    config_file_path_on_container, config_file_path_on_host = dump_container_config(
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/compress.py:73]: Might be a bit less confusing if we call these `generated_config_path_on_host` and `generated_config_path_on_container`.

File: components/clp-package-utils/clp_package_utils/scripts/decompress.py [hunk 4]
```diff
@@ -76,33 +77,15 @@ def main(argv):
...
-        "-u", f"{os.getuid()}:{os.getgid()}",
-        "--name", container_name,
-        "--log-driver", "local",
-        "--mount", str(mounts.clp_home),
-    ]
-    # fmt: on
-
+    config_file_path_on_container, config_file_path_on_host = dump_container_config(
+        clp_config, container_clp_config, container_name
+    )
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/decompress.py:84]:
```suggestion
    )

```

File: components/clp-package-utils/clp_package_utils/scripts/native/utils.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        # Create job
+        db_cursor.execute(
+            f"INSERT INTO `{QUERY_JOBS_TABLE_NAME}` (`job_config`, `type`) VALUES (%s, %s)",
+            (msgpack.packb(job_config.dict()), job_type),
+        )
+        db_conn.commit()
+        return db_cursor.lastrowid
+
+
+def wait_for_query_job(sql_adapter: SQL_Adapter, job_id: int) -> QueryJobStatus:
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/utils.py:65]:
```suggestion
def wait_for_query_job(sql_adapter: SQL_Adapter, job_id: int) -> QueryJobStatus:
    """
    Waits for the query job with the given ID to complete.
    :param sql_adapter: 
    :param job_id: 
    :return: The job's status on completion.
    """
```

File: components/clp-package-utils/clp_package_utils/scripts/native/utils.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    except asyncio.CancelledError:
+        pass
+    finally:
+        pool.terminate()
+        pool.close()
+
+
+def submit_query_job(
+    sql_adapter: SQL_Adapter, job_config: QueryJobConfig, job_type: QueryJobType
+) -> int:
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/utils.py:52]:
```suggestion
) -> int:
    """
    Submits a query job.
    :param sql_adapter:
    :param job_config:
    :param job_type:
    :return: The job's ID.
    """
```

File: components/clp-package-utils/clp_package_utils/general.py [hunk 7]
```diff
@@ -91,6 +101,10 @@ def get_clp_home():
     return clp_home.resolve()
 
 
+def generate_container_name(job_type: JobType) -> str:
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/general.py:104]:
```suggestion
def generate_container_name(job_type: JobType) -> str:
    """
    :param job_type:
    :return: A unique container name for the given job type.
    """
```

File: components/clp-package-utils/clp_package_utils/general.py [hunk 8]
```diff
@@ -177,7 +191,9 @@ def is_path_already_mounted(
     return host_path_relative_to_mounted_root == container_path_relative_to_mounted_root
 
 
-def generate_container_config(clp_config: CLPConfig, clp_home: pathlib.Path):
+def generate_container_config(
+    clp_config: CLPConfig, clp_home: pathlib.Path
+) -> Tuple[CLPConfig, CLPDockerMounts]:
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/general.py:196]: Let's also add the return value description:

```
    :return: The container config and the mounts.
```

File: components/clp-package-utils/clp_package_utils/general.py [hunk 9]
```diff
@@ -241,6 +257,43 @@ def generate_container_config(clp_config: CLPConfig, clp_home: pathlib.Path):
     return container_clp_config, docker_mounts
 
 
+def dump_container_config(
+    clp_config: CLPConfig, container_clp_config: CLPConfig, container_name: str
+) -> Tuple[pathlib.Path, pathlib.Path]:
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/general.py:262]:
```suggestion
) -> Tuple[pathlib.Path, pathlib.Path]:
    """
    Writes the given config to the logs directory so that it's accessible in the container.
    :param clp_config: The corresponding config on the host (used to determine the logs directory).
    :param container_clp_config: The config to write.
    :param container_name:
    :return: The path to the config file in the container and on the host.
    """
```

File: components/clp-package-utils/clp_package_utils/general.py [hunk 10]
```diff
@@ -241,6 +257,43 @@ def generate_container_config(clp_config: CLPConfig, clp_home: pathlib.Path):
...
+    config_file_path_on_container = container_clp_config.logs_directory / container_config_filename
+    with open(config_file_path_on_host, "w") as f:
+        yaml.safe_dump(container_clp_config.dump_to_primitive_dict(), f)
+
+    return config_file_path_on_container, config_file_path_on_host
+
+
+def generate_container_start_cmd(
+    container_name: str, container_mounts: List[CLPDockerMounts], execution_container: str
+) -> List[str]:
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/general.py:274]:
```suggestion
) -> List[str]:
    """
    Generates the command to start a container with the given mounts and name.
    :param container_name:
    :param container_mounts:
    :param execution_container:
    :return: The command.
    """
```

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py [hunk 11]
```diff
@@ -93,16 +104,6 @@ def main(argv):
...
-    if len(parsed_args.paths) > 0 and parsed_args.files_from is not None:
-        args_parser.error("Paths cannot be specified both on the command line and through a file.")
-
-    # Validate extraction directory
-    extraction_dir = pathlib.Path(parsed_args.extraction_dir)
-    if not extraction_dir.is_dir():
-        logger.error(f"extraction-dir ({extraction_dir}) is not a valid directory.")
-        return -1
-
     # Validate and load config file
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:107]: Can we move this into a method and then call it after validating the arguments? I think it's more intuitive to validate the arguments first, then the config file.

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py [hunk 12]
```diff
@@ -25,15 +25,24 @@
 logger.addHandler(logging_console_handler)
 
 
-def decompress_paths(
+def handle_decompression(
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:28]: In the next PR, can we rename this to `handle_decompression_cmd`?

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py [hunk 13]
```diff
@@ -46,6 +55,8 @@ def decompress_paths(
         "--db-config-file", str(db_config_file_path),
     ]
     # fmt: on
+    paths = parsed_args.paths
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:58]: Can we move this up, just after validating the args?

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py [hunk 14]
```diff
@@ -25,15 +25,24 @@
...
+def handle_decompression(
+    parsed_args,
     clp_home: pathlib.Path,
-    paths,
-    list_path: pathlib.Path,
     clp_config: CLPConfig,
-    archives_dir: pathlib.Path,
-    logs_dir: pathlib.Path,
-    extraction_dir: pathlib.Path,
 ):
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:32]:
```suggestion
):
    """
    Handles the decompression command.
    :param parsed_args:
    :param clp_home:
    :param clp_config:
    :return: 0 on success, -1 otherwise.
    """
```

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py [hunk 15]
```diff
@@ -25,15 +25,24 @@
...
+    # Validate paths were specified using only one method
+    if len(parsed_args.paths) > 0 and parsed_args.files_from is not None:
+        logger.error("Paths cannot be specified both on the command line and through a file.")
```
Comment (kirkrodrigues) [inline components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:35]:
```suggestion
        logger.error("Paths cannot be specified both on the command line and through a file.")
        return -1
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Thanks for doing this refactoring. Added some suggestions.

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py (hunk 11 above)
Comment (haiqi96) [reply components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:115]: like `validate_decompression_(command)_arguments`?

File: components/clp-package-utils/clp_package_utils/general.py (hunk 1 above)
Comment (haiqi96) [reply components/clp-package-utils/clp_package_utils/general.py:265]: technically, we don't need the `clp_config`. the only thing needed by this function is `clp_config.logs_directory` which can be an argument called `host_log_directory`.
I didn't write in this way because then `clp_config.logs_directory` need to be duplicated in all callers.
me know what you think

Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/general.py:265]: Since `clp_config.logs_directory` and `container_clp_config.logs_directory` are linked, I figure it's safer to pass in `container_clp_config` and `clp_config` to lessen the risk that the caller passes in two random paths.

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py (hunk 11 above)
Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:115]: Sorry, I mean can we move the `Validate and load config file` block.

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py (hunk 12 above)
Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:28]: Sure

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py [hunk 16]
```diff
@@ -25,15 +26,64 @@
 logger.addHandler(logging_console_handler)
 
 
-def decompress_paths(
+def _validate_and_load_config_file(
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:29]: can't call it `validate_and_load_config_file` because the function in `general.py` already has this name.

File: components/clp-package-utils/clp_package_utils/scripts/search.py [hunk 17]
```diff
@@ -76,46 +79,28 @@ def main(argv):
         clp_config.validate_logs_dir()
 
         # Validate and load necessary credentials
-        validate_and_load_db_credentials_file(clp_config, clp_home, True)
+        validate_and_load_db_credentials_file(clp_config, clp_home, False)
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/search.py:82]: Looks like a bug to me. Technically we can also factor out this `validate and load config block` into a common function and the submitter scripts (not the ones in native) can share it.  Just need to find a different name.

File: components/clp-package-utils/clp_package_utils/scripts/native/decompress.py (hunk 16 above)
Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/native/decompress.py:29]: How about renaming the existing `validate_and_load_config_file` to `load_config_file`?

File: components/clp-package-utils/clp_package_utils/scripts/search.py (hunk 17 above)
Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/search.py:82]: Makes sense

File: components/clp-package-utils/clp_package_utils/general.py [hunk 18]
```diff
@@ -249,9 +322,29 @@ def validate_config_key_existence(config, key):
     return value
 
 
-def validate_and_load_config_file(
+def validate_and_load_config_for_job_submission_script(
```
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/general.py:325]: can't think of a good name for this method. I don't want to call it `validate_and_load_config` because the usage of this method is only limited to the caller script under `package_utils/scripts`. I also feel it's ok to revert this change.

Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/general.py:325]: Yeah, maybe it's better to revert cause I'm also noticing that the different instances of this code validate different directories.

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

clp-package: Deduplicate and clean up code in CLI job-launcher scripts.
